CONTRIBUTING.md

```diff
@@ -0,0 +1,408 @@
+## Documentation style
+
+Documentation uses standard Doxygen.
+
+###  Developer targeted notes
+Make use of `@internal` doxygen tag. This can be used for any comment which is not intended for users, but is useful to better understand a piece of code.
+
+### Whitespaces
+
+#### After named tags such as `@param foo`
+The documentation tags which use an additional name should be followed by two spaces in order to better distinguish the text from the doxygen tag. It is also possible to use a line break instead.
+
+### Documenting examples
+There are two main steps:
+
+1. First, you can just copy over the [`doc/`](https://github.com/ginkgo-project/ginkgo/tree/develop/examples/simple-solver) folder (you can copy it from the example most relevant to you) and adapt your example names and such, then you can modify the actual documentation.  
++ In `tooltip`: A short description of the example.
++ In `short-intro`: The name of the example.
++ In `results.dox`: Run the example and write the output you get.
++ In `kind`: The kind of the example. For different kinds see [the documentation](https://ginkgo-project.github.io/ginkgo/doc/develop/Examples.html). Examples can be of `basic`, `techniques`, `logging`, `stopping_criteria` or `preconditioners`. If your example does not fit any of these categories, feel free to create one.
++ In `intro.dox`: You write an explanation of your code with some introduction similar to what you see in an existing example most relevant to you.
++ In `builds-on`: You write the examples it builds on. 
+
+2. You also need to modify the [examples.hpp.in](https://github.com/ginkgo-project/ginkgo/blob/develop/doc/examples/examples.hpp.in) file. You add the name of the example in the main section and in the section that you specified in the `doc/kind` file in the example documentation.
+
+
+## Other programming comments
+
+### C++ standard stream objects
+
+These are global objects and are shared inside the same translation unit. Therefore, whenever its state or formatting is changed (e.g. using `std::hex` or floating point formatting) inside library code, make sure to restore the state before returning the control to the user. See this [stackoverflow question](https://stackoverflow.com/questions/2273330/restore-the-state-of-stdcout-after-manipulating-it) for examples on how to do it correctly. This is extremely important for header files.
+
+### Warnings
+
+By default, the `-DGINKGO_COMPILER_FLAGS` is set to `-Wpedantic` and hence pedantic warnings are emitted by default. Some of these warnings are false positives and a complete list of the currently known warnings and their solutions is listed in [Issue 174](https://github.com/ginkgo-project/ginkgo/issues/174). Specifically, when macros are being used, we have the issue of having `extra ;` warnings, which is resolved by adding a `static_assert()`. The CI system additionally also has a step where it compiles for pedantic warnings to be errors.
+
+### Avoiding circular dependencies
+
```

Comment:
nit: You are a bit inconsistent with the blank lines in this file.
You always have at least one blank line before any heading, but sometimes there are 2, sometimes just 1 (I don't see a pattern when you do which). Maybe use 2 when you go from a smaller to a larger heading (e.g. from `###` to `##`), and 1 otherwise.
After a heading, you mostly use a single blank line, but sometimes, you don't have any.
I would suggest one of these rules:
- Always a blank line after a heading
- Only a blank line after a headding if there is more than 1 line of content.